File: 2015/03/1.py
input = None
try:
    input = open("./2015/03/input.txt", "r")
except:
    print("Error opening input.txt")
    exit(1)


class Coordinate:
    x = 0
    y = 0

    def __init__(self) -> None:
        self.x = 0
        self.y = 0

    def __str__(self) -> str:
        return "({},{})".format(self.x, self.y)


santaPos = Coordinate()
housesWithGifts = set()
housesWithGifts.add(str(santaPos))
for char in input.read():
    if char == "^":
        santaPos.y += 1
    elif char == "v":
        santaPos.y -= 1
    elif char == "<":
        santaPos.x -= 1
    elif char == ">":
        santaPos.x += 1
    else:
        print("Error, unknown character '", char, "'", sep="")
        exit(1)
    housesWithGifts.add(str(santaPos))

print(
    "Number of houses that recieve atleast one gift from santa is {}".format(
        len(housesWithGifts)
    )
)

File: 2015/03/1.go
package main

import (
	"fmt"
	"io"
	"os"
)

type Coordinate struct {
	x int
	y int
}

func main() {
	input, err := os.Open("./2015/03/input.txt")
	if err != nil {
		fmt.Fprintln(os.Stderr, "Unable to open input file")
		os.Exit(1)
	}
	var b = make([]byte, 1)
	santa := Coordinate{0, 0}
	houses := make(map[Coordinate]uint)
	houses[santa] = 1
	for true {
		_, err := input.Read(b)
		if err != nil {
			if err == io.EOF {
				break
			}
			fmt.Fprintln(os.Stderr, "Unexpected Error when reading input.txt", err)
			os.Exit(1)
		}
		switch string(b) {
		case "^":
			santa.y++
		case "v":
			santa.y--
		case "<":
			santa.x--
		case ">":
			santa.x++
		default:
			fmt.Fprintln(os.Stderr, "Unexpected character '", string(b), "'")
		}
		houses[santa]++
	}
	fmt.Println("Number of houses with atleast 1 gift is", len(houses))
}
